MayssaraOmar/assignment4/Counting islands.cpp

```diff
@@ -0,0 +1,125 @@
+#include "test.h"
+#include <iostream>
+#include <vector>
+using namespace std;
+
+static void markIsland(
+	const vector<vector<bool>> &grid,
+	int row,
+	int col,
+	vector<vector<bool>> *visited) {
+	// boundary check
+	if (row < 0 || row >= grid.size() || col < 0 || col >= grid[row].size()) {
+		return;
+	}
+	if ((*visited)[row][col] || !grid[row][col]) {
+		return;
+	}
+	(*visited)[row][col] = true;
+	markIsland(grid, row - 1, col, visited); // up
+	markIsland(grid, row + 1, col, visited); // down
+	markIsland(grid, row, col - 1, visited); // left
+	markIsland(grid, row, col + 1, visited); // right
+}
+
+int getIslandsCount(const vector<vector<bool>> &grid) {
+	vector<vector<bool>> visited(grid.size());
+	for (int i = 0; i < grid.size(); i++) {
+		visited[i].resize(grid[i].size(), false);
+	}
+	int count = 0;
+	for (int r = 0; r < grid.size(); r++) {
+		for (int c = 0; c < grid[r].size(); c++) {
+			if (grid[r][c] && !visited[r][c]) {
+				count++;
+				markIsland(grid, r, c, &visited);
+			}
+		}
+	}
+	return count;
+}
+
+// tests
+void emptyGrid_test() {
+	vector<vector<bool>> grid;
+	EXPECT_EQ(0, getIslandsCount(grid));
+}
+
+void landGrid_test() {
+	vector<vector<bool>> grid = {{true, true, true},
```

Comment:
Minor comment, having a unction to convert `string` -> `vector<vector<bool>>` or `vector<string>` -> `vector<vector<bool>>` count make the test cases more readable.